File: app/src/androidTest/java/djordjeh/architecture/mvp/dagger/TestApplicationModule.java
package djordjeh.architecture.mvp.dagger;

import java.util.List;

import javax.inject.Singleton;

import androidx.annotation.NonNull;
import dagger.Module;
import dagger.Provides;
import djordjeh.architecture.mvp.data.model.Task;
import djordjeh.architecture.mvp.data.source.TaskDataSource;
import djordjeh.architecture.mvp.ui.SchedulersFacade;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

@Module
public class TestApplicationModule {

    private TaskDataSource taskDataSource;

    public TestApplicationModule(TaskDataSource taskDataSource) {
        this.taskDataSource = taskDataSource;
    }

    @Provides
    @Singleton
    TaskDataSource taskRepository() {
        return new MockTasksDataSource(taskDataSource);
    }

    @Provides
    @Singleton
    SchedulersFacade provideSchedulersFacade() {
        return new TestSchedulersFacadeImpl();
    }

    static class TestSchedulersFacadeImpl implements SchedulersFacade {

        @Override
        public Scheduler io() {
            return Schedulers.trampoline();
        }

        @Override
        public Scheduler computation() {
            return Schedulers.trampoline();
        }

        @Override
        public Scheduler mainThread() {
            return Schedulers.trampoline();
        }
    }

    static class MockTasksDataSource implements TaskDataSource {

        private TaskDataSource taskDataSource;

        MockTasksDataSource(TaskDataSource taskDataSource) {
            this.taskDataSource = taskDataSource;
        }

        @Override
        public Observable<List<Task>> tasks(boolean forceUpdate) {
            return taskDataSource.tasks(forceUpdate);
        }

        @Override
        public Maybe<Task> task(long taskId) {
            return taskDataSource.task(taskId);
        }

        @Override
        public Single<Task> save(@NonNull Task task) {
            return Single.just(task);
        }

        @Override
        public Single<Task> delete(@NonNull Task task) {
            return Single.just(task);
        }
    }
}

File: app/src/androidTest/java/djordjeh/architecture/mvp/dagger/TestApplicationModule.kt
package djordjeh.architecture.mvp.dagger

import javax.inject.Singleton
import dagger.Module
import dagger.Provides
import djordjeh.architecture.mvp.data.model.Task
import djordjeh.architecture.mvp.data.source.TaskDataSource
import djordjeh.architecture.mvp.ui.SchedulersFacade
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

@Module
class TestApplicationModule(private val taskDataSource: TaskDataSource) {

    @Provides
    @Singleton
    internal fun taskRepository(): TaskDataSource {
        return MockTasksDataSource(taskDataSource)
    }

    @Provides
    @Singleton
    internal fun provideSchedulersFacade(): SchedulersFacade {
        return TestSchedulersFacadeImpl()
    }

    internal class TestSchedulersFacadeImpl : SchedulersFacade {

        override fun io(): Scheduler {
            return Schedulers.trampoline()
        }

        override fun computation(): Scheduler {
            return Schedulers.trampoline()
        }

        override fun mainThread(): Scheduler {
            return Schedulers.trampoline()
        }
    }

    internal class MockTasksDataSource(private val taskDataSource: TaskDataSource) : TaskDataSource {

        override fun tasks(forceUpdate: Boolean): Observable<List<Task>> {
            return taskDataSource.tasks(forceUpdate)
        }

        override fun task(taskId: Long): Maybe<Task> {
            return taskDataSource.task(taskId)
        }

        override fun save(task: Task): Single<Task> {
            return Single.just(task)
        }

        override fun delete(task: Task): Single<Task> {
            return Single.just(task)
        }
    }
}
